v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/compilation-job.ts

```diff
@@ -108,20 +137,58 @@ export class CompilationJobImplementation implements CompilationJob {
     };
   }
 
-  #computeBuildId(): string {
+  async #getSourceContentHash(file: ResolvedFile): Promise<string> {
+    const cachedSourceContentHash =
+      CompilationJobImplementation.#sourceContentHashCache.get(file.sourceName);
+    if (cachedSourceContentHash !== undefined) {
+      return cachedSourceContentHash;
+    }
+
+    const sourceContentHash = await createNonCryptographicHashId(
+      file.content.text,
+    );
+    CompilationJobImplementation.#sourceContentHashCache.set(
+      file.sourceName,
+      sourceContentHash,
+    );
+    return sourceContentHash;
+  }
+
+  async #computeBuildId(): Promise<string> {
     // NOTE: We type it this way so that this stop compiling if we ever change
     // the format of the BuildInfo type.
     const format: BuildInfo["_format"] = "hh3-sol-build-info-1";
 
+    const sources: { [sourceName: string]: { hash: string } } = {};
+    const resolvedFiles = this.#getResolvedFiles();
+
+    await Promise.all(
+      resolvedFiles.map(async (file) => {
+        sources[file.sourceName] = {
+          hash: await this.#getSourceContentHash(file),
+        };
+      }),
+    );
+
+    const sortedSources = Object.fromEntries(
+      Object.entries(sources).sort((a, b) => a[0].localeCompare(b[0])),
+    );
```

Comment:
We need to sort the sources because the sources map might be populated out of order which does affect serialisation.

Reply:
great point. What about adding a comment?

Reply:
Done :D